### PR TITLE
bpo-33234 Improve list() pre-sizing for inputs with known lengths (no __length_hint__) 

### PR DESCRIPTION
**Pull request by pablogsal**
This is a version of https://github.com/python/cpython/pull/6493 that only preallocates the list when `__len__` is available to reduce memory overallocation and possible performance hit on big iterables.

This are the L1 and L2 cache misses and related information:

THIS PR
```
❯ perf stat -r 200 -B -e cache-references,cache-misses,cycles,instructions,branches,faults,migrations ./python -c "
for _ in range(1000):
    list([0]*10000)
"
 Performance counter stats for './python -c
for _ in range(1000):
    list([0]*10000)
' (200 runs):

         1,151,365      cache-references:u                                            ( +-  0.47% )
             5,686      cache-misses:u            #    0.494 % of all cache refs      ( +-  8.06% )
       370,237,437      cycles:u                                                      ( +-  0.05% )
       542,049,270      instructions:u            #    1.46  insn per cycle           ( +-  0.00% )
       134,146,925      branches:u                                                    ( +-  0.00% )
             8,031      faults:u                                                      ( +-  0.00% )
                 0      migrations:u

          0.122170 +- 0.000307 seconds time elapsed  ( +-  0.25% )
```

MASTER
```
❯ perf stat -r 200 -B -e cache-references,cache-misses,cycles,instructions,branches,faults,migrations ./python -c "for _ in range(1000):
    list([0]*10000)
"
 Performance counter stats for './python -c
for _ in range(1000):
    list([0]*10000)
' (200 runs):

         1,275,440      cache-references:u                                            ( +-  0.61% )
            21,895      cache-misses:u            #    1.717 % of all cache refs      ( +-  4.69% )
       392,293,096      cycles:u                                                      ( +-  0.22% )
       543,254,939      instructions:u            #    1.38  insn per cycle           ( +-  0.00% )
       134,549,173      branches:u                                                    ( +-  0.00% )
            10,039      faults:u                                                      ( +-  0.00% )
                 0      migrations:u

          0.133171 +- 0.000463 seconds time elapsed  ( +-  0.35% )
```

https://bugs.python.org/issue33234

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2018-04-17-01-24-51.bpo-33234.l9IDtp.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2018-04-17-01-24-51.bpo-33234.l9IDtp.rst]: length, not size.

File: Misc/NEWS.d/next/Core and Builtins/2018-04-17-01-24-51.bpo-33234.l9IDtp.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2018-04-17-01-24-51.bpo-33234.l9IDtp.rst]: "or can be reasonably estimated" I don't understand this part.

File: Objects/listobject.c
Comment (vstinner) [inline Objects/listobject.c]: nitpick: ...) { (add a space)

**Review by vstinner — APPROVED** [review]
LGTM, but the NEWS entry should be fixed.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
I suggest to wait the approval of a second core dev before merging this change.

**Comment by pablogsal** [comment]
I would like @serhiy-storchaka comment and advice on this if possible.

**Comment by vstinner** [comment]
@pablogsal: If you want me to merge your PR, I would be me confident if you write a benchmark for the worst case: a small container (ex: 10 items or less) without a known length. I expect that the additional function call (obj.__len__()) would slow down this case. I would be interested to see the overhead.

**Comment by pablogsal** [comment]
@vstinner 

Experiment:
```python
import perf

runner = perf.Runner()
runner.timeit("list(x)",
              stmt="list(x)",
              setup="x = [0]*10000")
```

result:
```
 ./python.exe -m perf  compare_to old.json new.json
Mean +- std dev: [old] 795 ns +- 25 ns -> [new] 2.94 us +- 0.07 us: 3.70x slower (+270%)
```

**Comment by pablogsal** [comment]
This is the first experiment using `PyObject_LengthHint`:

```
❯ ./python.exe -m perf  compare_to ../cpython/old.json new.json
Mean +- std dev: [old] 795 ns +- 25 ns -> [new] 895 ns +- 31 ns: 1.13x slower (+13%)```

**Comment by vstinner** [comment]
> Mean +- std dev: [old] 795 ns +- 25 ns -> [new] 2.94 us +- 0.07 us: 3.70x slower (+270%)

This overhead likely comes from the raisin an exception + PyErr_Clear(). hasattr() has been optimized by adding a _PyObject_LookupAttr() which is getattr() without raising an exception if the attribute doesn't exist.

Probing an object size is a common need for optimization to preallocate a container. Python builtin types would benefit of such function.

What do you think of adding something like _PyObject_ProbeLength() which is like PyObject_Length() without raising an exception?

cc @methane @serhiy-storchaka

File: Objects/listobject.c
Comment (vstinner) [inline Objects/listobject.c]: To reduce or avoid the overhead of the worst case, you can avoid this call if _PyObject_HasLen() is false.

This change doesn't require to implement a new _PyObject_ProbeLength().

Comment (pablogsal) [reply Objects/listobject.c]: Updated run of first experiment using `_PyObject_HasLen()`:

```
Mean +- std dev: [old] 795 ns +- 25 ns -> [new] 776 ns +- 25 ns: 1.02x faster (-6%)
```

**Comment by pablogsal** [comment]
Updated run of first experiment using `_PyObject_HasLen()`:

```
Mean +- std dev: [old] 795 ns +- 25 ns -> [new] 776 ns +- 25 ns: 1.02x faster (-6%)
```

File: Objects/listobject.c
Comment (serhiy-storchaka) [inline Objects/listobject.c]: Add a space before `{`.

File: Objects/listobject.c
Comment (serhiy-storchaka) [inline Objects/listobject.c:2685]: I'm not comfortable with silencing all exceptions. It can be a KeyboardInterrupt for example.

File: Objects/listobject.c
Comment (serhiy-storchaka) [inline Objects/listobject.c]: `PyObject_Length` is an alias of `PyObject_Size` (for backward compatibility). It is better to use `PyObject_Size`.

File: Objects/listobject.c
Comment (serhiy-storchaka) [inline Objects/listobject.c:101]: It is not guarantied that ob_item is NULL here. New items can be added in destructors called in `_list_clear()`.

File: Objects/listobject.c
Comment (serhiy-storchaka) [inline Objects/listobject.c]: Would not be easier to use `PyMem_New`?

File: Objects/listobject.c
Comment (pablogsal) [reply Objects/listobject.c:101]: In b678442 I modified the enter condition for the new preallocation so it goes directly to the `list_extend` path if there are objects after calling `_list_clear(self)`:

```C
if (_PyObject_HasLen(iterable) && self->ob_item == NULL) {
```

**Comment by pablogsal** [comment]
Performance metrics with commit b678442:

❯ ./python -m perf compare_to ../cpython2/old.json new.json
Mean +- std dev: [old_big] 97.5 us +- 3.4 us -> [new_big] 88.6 us +- 5.5 us: 1.1x faster (-5%)

**Review by methane — APPROVED** [review]
LGTM

File: Objects/listobject.c
Comment (vstinner) [reply Objects/listobject.c:101]: Well, it wouldn't hurt to add assert(self->ob_item == NULL) at this function start. I see that as documentation of the function contract ;-)

File: Objects/listobject.c
Comment (vstinner) [inline Objects/listobject.c]: nitpick: PEP 7 requires

```
}
else {
```

**Review by vstinner — APPROVED** [review]
LGTM, it's now even better with _PyObject_HasLen()!

File: Objects/listobject.c
Comment (serhiy-storchaka) [inline Objects/listobject.c]: What tests will fail if not silence any exceptions at all?

File: Objects/listobject.c
Comment (serhiy-storchaka) [reply Objects/listobject.c:101]: `PyObject_Size()` can call the Python code and release GIL. This can lead to adding items to the list.

File: Objects/listobject.c
Comment (pablogsal) [reply Objects/listobject.c]: At least the following:

```traceback
test_generators failed
**********************************************************************
File "/home/pablogsal/github/cpython/Lib/test/test_generators.py", line ?, in test.test_generators.__test__.weakref
Failed example:
    list(p)
Exception raised:
    Traceback (most recent call last):
      File "/home/pablogsal/github/cpython/Lib/doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest test.test_generators.__test__.weakref[9]>", line 1, in <module>
        list(p)
    TypeError: object of type 'generator' has no len()
**********************************************************************
1 items had failures:
   1 of  10 in test.test_generators.__test__.weakref
***Test Failed*** 1 failures.


test_genexps failed
**********************************************************************
File "/home/pablogsal/github/cpython/Lib/test/test_genexps.py", line ?, in test.test_genexps.__test__.doctests
Failed example:
    list(p)
Exception raised:
    Traceback (most recent call last):
      File "/home/pablogsal/github/cpython/Lib/doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest test.test_genexps.__test__.doctests[75]>", line 1, in <module>
        list(p)
    TypeError: object of type 'generator' has no len()
**********************************************************************
1 items had failures:
   1 of  76 in test.test_genexps.__test__.doctests
***Test Failed*** 1 failures.

```

File: Objects/listobject.c
Comment (pablogsal) [reply Objects/listobject.c:101]: Addressed in b6ff7ce

File: Objects/listobject.c
Comment (vstinner) [inline Objects/listobject.c]: nitpick: for readability, I suggest to invert the if condition and exchange the true/false blocks, and move PyErr_Clear() after the if:

if (!not exception) { return -1; }
PyErr_Clear();

**Review by vstinner — APPROVED** [review]
LGTM.

File: Objects/listobject.c
Comment (serhiy-storchaka) [reply Objects/listobject.c]: How can `_PyObject_HasLen()` return true for generators?

Comment (serhiy-storchaka) [reply Objects/listobject.c]: Ah, it is not true generator, it is a weakref. What if silence just TypeError?

Comment (pablogsal) [reply Objects/listobject.c]: I am a bit afraid if someone has similar tests/usage like these two cases and then they will receive new errors as raising all exceptions would be technically a backwards incompatible change. But if you think we should raise (maybe ignoring `TypeError`) I am happy to change the implementation.

Comment (serhiy-storchaka) [reply Objects/listobject.c]: See `PyObject_LengthHint()`.

Comment (pablogsal) [reply Objects/listobject.c]: Done in 0a6c8ba

File: Objects/listobject.c
Comment (vstinner) [inline Objects/listobject.c:2685]: @pablogsal, @serhiy-storchaka: In a previous comment, I proposed to add an helper function to "probe" an object size: so move this code into a private helper function. Since the same code is used by PyObject_LengthHint(), it would now make sense, no?

See also iter_len() ... which is different.

Comment (serhiy-storchaka) [reply Objects/listobject.c:2685]: This code will be used just in two places and is not too complex. Adding yet one intermediate function will add a performance penalty for calling a function and checking its result, and will complicate the code. If this code will be used in more places, it can be refactored.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**
